[Messenger] Remove indices in messenger table on MySQL to prevent deadlocks while removing messages when running multiple consumers

jeroennoten opened the pull request:
SELECT ... FOR UPDATE locks rows but also relevant indices. Since locking rows and indices is not one atomic operation,
this might cause deadlocks when running multiple workers. Removing indices on queue_name and available_at
resolves this problem.

| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | Fix #41541 #39041
| License       | MIT

Using Doctrine transport with multiple consumers occasionally results in MySQL deadlocks while removing a message from the messages database table.

This can be reproduced consistently by setting up a default `async` queue with the Doctrine transport and creating an empty `TestMessage` and `TestMessageHandler`. Create a command that dispatches 10000 of these messages in a for loop en start 4 message consumers. After a while, several consumers report a deadlock:

```
In Connection.php line 227:
                                                                                                                   
  [Symfony\Component\Messenger\Exception\TransportException]                                                       
  An exception occurred while executing 'DELETE FROM messenger_messages WHERE id = ?' with params ["32903"]:       
                                                                                                                   
  SQLSTATE[40001]: Serialization failure: 1213 Deadlock found when trying to get lock; try restarting transaction  
                                                                                                                   

Exception trace:
  at /var/www/vendor/symfony/messenger/Transport/Doctrine/Connection.php:227
 Symfony\Component\Messenger\Transport\Doctrine\Connection->ack() at /var/www/vendor/symfony/messenger/Transport/Doctrine/DoctrineReceiver.php:79
 Symfony\Component\Messenger\Transport\Doctrine\DoctrineReceiver->ack() at /var/www/vendor/symfony/messenger/Transport/Doctrine/DoctrineTransport.php:50
 Symfony\Component\Messenger\Transport\Doctrine\DoctrineTransport->ack() at /var/www/vendor/symfony/messenger/Worker.php:150
 Symfony\Component\Messenger\Worker->handleMessage() at /var/www/vendor/symfony/messenger/Worker.php:81
 Symfony\Component\Messenger\Worker->run() at /var/www/vendor/symfony/messenger/Command/ConsumeMessagesCommand.php:202
 Symfony\Component\Messenger\Command\ConsumeMessagesCommand->execute() at /var/www/vendor/symfony/console/Command/Command.php:255
 Symfony\Component\Console\Command\Command->run() at /var/www/vendor/symfony/console/Application.php:1027
 Symfony\Component\Console\Application->doRunCommand() at /var/www/vendor/symfony/framework-bundle/Console/Application.php:97
 Symfony\Bundle\FrameworkBundle\Console\Application->doRunCommand() at /var/www/vendor/symfony/console/Application.php:273
 Symfony\Component\Console\Application->doRun() at /var/www/vendor/symfony/framework-bundle/Console/Application.php:83
 Symfony\Bundle\FrameworkBundle\Console\Application->doRun() at /var/www/vendor/symfony/console/Application.php:149
 Symfony\Component\Console\Application->run() at /var/www/bin/console:42
```

A similar problem with Laravel's queue worker (and a solution) is reported here: https://github.com/laravel/framework/issues/31660

The solution is to remove indices on the `queue_name` and `available_at` columns. After removing these indices, I could not reproduce the issue anymore. Also, I did not notice any performance degradations.

jeroennoten committed:
Remove indices in messenger table on MySQL to prevent deadlocks while removing messages when running multiple consumers

SELECT ... FOR UPDATE row locks also locks indices. Since locking rows and indices is not one atomic operation,
this might cause deadlocks when running multiple workers. Removing indices on queue_name and available_at
resolves this problem.

fabpot commented:
IIUC, this issue only exists on MySQL. It this is the case, the change should only be done for MySQL and not for all database engines.

jeroennoten commented:
@fabpot, you are right. To be sure, I tested with PostgreSQL and I could not reproduce the issues there.

So I changed this to only apply to MySQL. I also added a unit test for these changes.

fabpot commented on src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php:
```diff
@@ -386,7 +387,6 @@ private function getSchema(): Schema
         $table->addColumn('headers', self::$useDeprecatedConstants ? Type::TEXT : Types::TEXT)
             ->setNotnull(true);
         $table->addColumn('queue_name', self::$useDeprecatedConstants ? Type::STRING : Types::STRING)
-            ->setLength(190) // MySQL 5.6 only supports 191 characters on an indexed column in utf8mb4 mode
```
This should be reverted as we are not creating an index on MySQL for that column anymore, right?

jeroennoten replied:
Right, so that is exactly why I removed that line.

fabpot commented:
Thank you @jeroennoten.

bobvandevijver commented:
Thank you @jeroennoten, I just saw this exact error in production for the very first time this weekend. Very nice to see it has already been resolved!

jeroennoten commented:
@bobvandevijver, you're welcome!

Don't forget to run `bin/console messenger:setup-transports` after updating Symfony to the latest version to apply these changes to the database. The command will update the existing table or create a new one if it does not exist.

bobvandevijver commented:
We're using migrations to define the table, but thanks for the tip!

palttamas commented:
@jeroennoten removing the indexes fixes the lock issue but caused us significant performance issues. I was searching the issues and it seems to me that multiple consumers are not working very well for others either, be it redis or doctrine transport, no matter.

nicolas-grekas commented:
I'm proposing to revert this PR and replace it by another approach in #45888
